# Assignment 5(Implementing TFIDF vectorizer)

<font face='georgia'>
    
   <h4><strong>What does tf-idf mean?</strong></h4>

   <p>    
Tf-idf stands for <em>term frequency-inverse document frequency</em>, and the tf-idf weight is a weight often used in information retrieval and text mining. This weight is a statistical measure used to evaluate how important a word is to a document in a collection or corpus. The importance increases proportionally to the number of times a word appears in the document but is offset by the frequency of the word in the corpus. Variations of the tf-idf weighting scheme are often used by search engines as a central tool in scoring and ranking a document's relevance given a user query.
</p>
    
   <p>
One of the simplest ranking functions is computed by summing the tf-idf for each query term; many more sophisticated ranking functions are variants of this simple model.
</p>
    
   <p>
Tf-idf can be successfully used for stop-words filtering in various subject fields including text summarization and classification.
</p>
    
</font>

<font face='georgia'>
    <h4><strong>How to Compute:</strong></h4>

Typically, the tf-idf weight is composed by two terms: the first computes the normalized Term Frequency (TF), aka. the number of times a word appears in a document, divided by the total number of words in that document; the second term is the Inverse Document Frequency (IDF), computed as the logarithm of the number of the documents in the corpus divided by the number of documents where the specific term appears.

 <ul>
    <li>
<strong>TF:</strong> Term Frequency, which measures how frequently a term occurs in a document. Since every document is different in length, it is possible that a term would appear much more times in long documents than shorter ones. Thus, the term frequency is often divided by the document length (aka. the total number of terms in the document) as a way of normalization: <br>

$TF(t) = \frac{\text{Number of times term t appears in a document}}{\text{Total number of terms in the document}}.$
</li>
<li>
<strong>IDF:</strong> Inverse Document Frequency, which measures how important a term is. While computing TF, all terms are considered equally important. However it is known that certain terms, such as "is", "of", and "that", may appear a lot of times but have little importance. Thus we need to weigh down the frequent terms while scale up the rare ones, by computing the following: <br>

$IDF(t) = \log_{e}\frac{\text{Total  number of documents}} {\text{Number of documents with term t in it}}.$
for numerical stabiltiy we will be changing this formula little bit
$IDF(t) = \log_{e}\frac{\text{Total  number of documents}} {\text{Number of documents with term t in it}+1}.$
</li>
</ul>

<br>
<h4><strong>Example</strong></h4>
<p>

Consider a document containing 100 words wherein the word cat appears 3 times. The term frequency (i.e., tf) for cat is then (3 / 100) = 0.03. Now, assume we have 10 million documents and the word cat appears in one thousand of these. Then, the inverse document frequency (i.e., idf) is calculated as log(10,000,000 / 1,000) = 4. Thus, the Tf-idf weight is the product of these quantities: 0.03 * 4 = 0.12.
</p>
</font>

## Task-1

<font face='georgia'>
    <h4><strong>1. Build a TFIDF Vectorizer & compare its results with Sklearn:</strong></h4>

<ul>
    <li> As a part of this task you will be implementing TFIDF vectorizer on a collection of text documents.</li>
    <br>
    <li> You should compare the results of your own implementation of TFIDF vectorizer with that of sklearns implemenation TFIDF vectorizer.</li>
    <br>
    <li> Sklearn does few more tweaks in the implementation of its version of TFIDF vectorizer, so to replicate the exact results you would need to add following things to your custom implementation of tfidf vectorizer:
       <ol>
        <li> Sklearn has its vocabulary generated from idf sorted in alphabetical order</li>
        <li> Sklearn formula of idf is different from the standard textbook formula. Here the constant <strong>"1"</strong> is added to the numerator and denominator of the idf as if an extra document was seen containing every term in the collection exactly once, which prevents zero divisions.
            
 $IDF(t) = 1+\log_{e}\frac{1\text{ }+\text{ Total  number of documents in collection}} {1+\text{Number of documents with term t in it}}.$
        </li>
        <li> Sklearn applies L2-normalization on its output matrix.</li>
        <li> The final output of sklearn tfidf vectorizer is a sparse matrix.</li>
    </ol>
    <br>
    <li>Steps to approach this task:
    <ol>
        <li> You would have to write both fit and transform methods for your custom implementation of tfidf vectorizer.</li>
        <li> Print out the alphabetically sorted voacb after you fit your data and check if its the same as that of the feature names from sklearn tfidf vectorizer. </li>
        <li> Print out the idf values from your implementation and check if its the same as that of sklearns tfidf vectorizer idf values. </li>
        <li> Once you get your voacb and idf values to be same as that of sklearns implementation of tfidf vectorizer, proceed to the below steps. </li>
        <li> Make sure the output of your implementation is a sparse matrix. Before generating the final output, you need to normalize your sparse matrix using L2 normalization. You can refer to this link https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html </li>
        <li> After completing the above steps, print the output of your custom implementation and compare it with sklearns implementation of tfidf vectorizer.</li>
        <li> To check the output of a single document in your collection of documents,  you can convert the sparse matrix related only to that document into dense matrix and print it.</li>
        </ol>
    </li>
    <br>
   </ul>

  <p> <font color="#e60000"><strong>Note-1: </strong></font> All the necessary outputs of sklearns tfidf vectorizer have been provided as reference in this notebook, you can compare your outputs as mentioned in the above steps, with these outputs.<br>
   <font color="#e60000"><strong>Note-2: </strong></font> The output of your custom implementation and that of sklearns implementation would match only with the collection of document strings provided to you as reference in this notebook. It would not match for strings that contain capital letters or punctuations, etc, because sklearn version of tfidf vectorizer deals with such strings in a different way. To know further details about how sklearn tfidf vectorizer works with such string, you can always refer to its official documentation.<br>
   <font color="#e60000"><strong>Note-3: </strong></font> During this task, it would be helpful for you to debug the code you write with print statements wherever necessary. But when you are finally submitting the assignment, make sure your code is readable and try not to print things which are not part of this task.
    </p>

### Corpus

In [32]:
## SkLearn# Collection of string documents

corpus = [
     'this is the first document',
     'this document is the second document',
     'and this is the third one',
     'is this the first document',
]

### SkLearn Implementation

In [33]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer()
skl_output = vectorizer.fit_transform(corpus)
skl_output

<4x9 sparse matrix of type '<class 'numpy.float64'>'
	with 21 stored elements in Compressed Sparse Row format>

In [34]:
# sklearn feature names, they are sorted in alphabetic order by default.

print(vectorizer.get_feature_names())

['and', 'document', 'first', 'is', 'one', 'second', 'the', 'third', 'this']


In [35]:
# Here we will print the sklearn tfidf vectorizer idf values after applying the fit method
# After using the fit function on the corpus the vocab has 9 words in it, and each has its idf value.

print(vectorizer.idf_)

[1.91629073 1.22314355 1.51082562 1.         1.91629073 1.91629073
 1.         1.91629073 1.        ]


In [36]:
vectorizer.idf_

array([1.91629073, 1.22314355, 1.51082562, 1.        , 1.91629073,
       1.91629073, 1.        , 1.91629073, 1.        ])

In [37]:
# shape of sklearn tfidf vectorizer output after applying transform method.

skl_output.shape

(4, 9)

In [38]:
# sklearn tfidf values for first line of the above corpus.
# Here the output is a sparse matrix

print(skl_output)

  (0, 1)	0.46979138557992045
  (0, 2)	0.5802858236844359
  (0, 6)	0.38408524091481483
  (0, 3)	0.38408524091481483
  (0, 8)	0.38408524091481483
  (1, 5)	0.5386476208856763
  (1, 1)	0.6876235979836938
  (1, 6)	0.281088674033753
  (1, 3)	0.281088674033753
  (1, 8)	0.281088674033753
  (2, 4)	0.511848512707169
  (2, 7)	0.511848512707169
  (2, 0)	0.511848512707169
  (2, 6)	0.267103787642168
  (2, 3)	0.267103787642168
  (2, 8)	0.267103787642168
  (3, 1)	0.46979138557992045
  (3, 2)	0.5802858236844359
  (3, 6)	0.38408524091481483
  (3, 3)	0.38408524091481483
  (3, 8)	0.38408524091481483


In [39]:
skl_output.shape[1]

9

In [40]:
# sklearn tfidf values for first line of the above corpus.
# To understand the output better, here we are converting the sparse output matrix to dense matrix and printing it.
# Notice that this output is normalized using L2 normalization. sklearn does this by default.

print(skl_output.toarray())

[[0.         0.46979139 0.58028582 0.38408524 0.         0.
  0.38408524 0.         0.38408524]
 [0.         0.6876236  0.         0.28108867 0.         0.53864762
  0.28108867 0.         0.28108867]
 [0.51184851 0.         0.         0.26710379 0.51184851 0.
  0.26710379 0.51184851 0.26710379]
 [0.         0.46979139 0.58028582 0.38408524 0.         0.
  0.38408524 0.         0.38408524]]


### Your custom implementation

In [42]:
# Write your code here.
# Make sure its well documented and readble with appropriate comments.
# Compare your results with the above sklearn tfidf vectorizer
# You are not supposed to use any other library apart from the ones given below

from collections import Counter
from tqdm import tqdm
from scipy.sparse import csr_matrix
import math
import operator
from sklearn.preprocessing import normalize
import numpy as np

In [43]:
## SkLearn# Collection of string documents

corpus = [
     'this is the first document',
     'this document is the second document',
     'and this is the third one',
     'is this the first document',
]

## Fit Method 
### (for features)

In [44]:
#fit function defined to find the unique words in corpus
def fit(dataset):    
    unique_words = set()
    # check if its list type or not
    if isinstance(dataset, (list)): #the document corpus is to be accepted in form of a list of sentences
        for row in dataset:
            for word in row.split(" "):
                if len(word) < 2:
                    continue
                unique_words.add(word)
        unique_words = sorted(list(unique_words))#words are sorted alphabetically
        return unique_words        

In [45]:
vocab = fit(corpus) # this would be our features or tthe unique words array
vocab

['and', 'document', 'first', 'is', 'one', 'second', 'the', 'third', 'this']

using fit function we see that our words list match the sklearn feature names on tfidf vectorizer implementation.

##### finding idf values for the features


for numerical stabiltiy we will be changing the default formula little bit

 $IDF(t) = 1+\log_{e}\frac{1\text{ }+\text{ Total  number of documents in collection}} {1+\text{Number of documents with term t in it}}.$

In [46]:
def vectorizer_idf(corpus,vocab):
    idf_matrix = []
    for w in vocab:
        df = 0
        for doc in corpus:
            s = set(doc.split()) #to eliminate the duplicate word cound in a document
            for i in s:
                if i == w:
                    df+=1
                    idf = 1+(math.log((1+len(corpus))/(1+df))) #idf formula mentioned above
        idf_matrix.append(idf)
    return(idf_matrix)

In [47]:
np.array(vectorizer_idf(corpus,vocab)) #finding the idf values for our features array

array([1.91629073, 1.22314355, 1.51082562, 1.        , 1.91629073,
       1.91629073, 1.        , 1.91629073, 1.        ])

Here we see that the idf values match to that of the sklearn tf-idf implementation

### Now we'll define a transform function that provides the tfidf values taking our corpus & the features as input
#### (The output that our function gives is in form of a l2 normalized sparse matrix,as sklearn tfidf implementation)

## Transform method:

In [48]:
def transform(corpus,words):
    
    #the term-frequency calculation for the each term in document corpus
    tf_matrix=[] # to store term frequency values for each feature in document corpus
    for doc in corpus:
        termfreq = [];
        doclist = doc.split()
        for w in words:
            tf = doclist.count(w)/len(doclist)
            termfreq.append(tf)
        tf_matrix.append(termfreq)
    
    # inverse document frequency calculation for the features in document corpus
    idf_matrix = [] #to store idf values for each feature in document corpus
    for w in words:
        df = 0;idf = 0;
        for doc in corpus:
            s = set(doc.split())
            for i in s:
                if i == w:
                    df+=1
                    idf = 1+(math.log((1+len(corpus))/(1+df))) #idf formula for numerical stability
        idf_matrix.append(idf) 
        
    # calculate the tfidf values for our document corpus
    tfidf_matrix=[]
    for i in tqdm(range(len(tf_matrix)),desc="TF-iDF vectorizer"):
        a=0;tfidf_doc =[]
        for j in range(len(idf_matrix)):
            tfidf = tf_matrix[i][j]*idf_matrix[j] #using the tfidf formula (TFiDF = TF * iDF)
            tfidf_doc.append(tfidf)
        tfidf_matrix.append(tfidf_doc)


    tfx = normalize(tfidf_matrix,norm = 'l2') #applying l2 normalization to our tfidf values matrix
    tfidfx = tfx.tolist() #converting the array to list for further process
    
    #iterating to find the non-zero rows,columns & values in our above tfidf values list-tfidfx
    rows=[];columns = [];tfidf_vals=[] #initiating for rows,columns & data values for our sparse matrix
    for i in tqdm(range(len(tfidfx)),desc = "Sparse matrix"):
        for j in range(len(tfidfx[i])):
            if tfidfx[i][j] != 0.0:
                rows.append(i);columns.append(j)
                tfidf_vals.append(tfidfx[i][j]) #storing only the non-zero tfidf values for sparse matrix conversion

    row = np.array(rows);col = np.array(columns);data= np.array(tfidf_vals)#converting lists to arrays
    tfidf_SX = csr_matrix((data,(row, col)),shape = (len(tfidfx),len(words)))#csr_sparse matrix conversion(our final result)

    return(tfidf_SX)

In [49]:
#the given data corpus that contaons list of sentences
corpus = [
     'this is the first document',
     'this document is the second document',
     'and this is the third one',
     'is this the first document',]

In [50]:
vocab = fit(corpus) #fit to find the features
print("Feature names sorted in alphabetical order\n",vocab)
print(" \n ")
print("Tfidf values sparse matrix \n ")
print(transform(corpus,vocab)) #transform to find the tfidf values

Sparse matrix: 100%|█████████████████████████████████████████████████████████████████████████████| 4/4 [00:00<?, ?it/s]

Feature names sorted in alphabetical order
 ['and', 'document', 'first', 'is', 'one', 'second', 'the', 'third', 'this']
 
 
Tfidf values sparse matrix 
 
  (0, 1)	0.4697913855799205
  (0, 2)	0.580285823684436
  (0, 3)	0.3840852409148149
  (0, 6)	0.3840852409148149
  (0, 8)	0.3840852409148149
  (1, 1)	0.6876235979836937
  (1, 3)	0.2810886740337529
  (1, 5)	0.5386476208856762
  (1, 6)	0.2810886740337529
  (1, 8)	0.2810886740337529
  (2, 0)	0.511848512707169
  (2, 3)	0.267103787642168
  (2, 4)	0.511848512707169
  (2, 6)	0.267103787642168
  (2, 7)	0.511848512707169
  (2, 8)	0.267103787642168
  (3, 1)	0.4697913855799205
  (3, 2)	0.580285823684436
  (3, 3)	0.3840852409148149
  (3, 6)	0.3840852409148149
  (3, 8)	0.3840852409148149


#### On comparing we see that the features names & the tfidf values match the sklearn implementation

## Task-2

<font face='georgia'>
    <h4><strong>2. Implement max features functionality:</strong></h4>

<ul>
    <li> As a part of this task you have to modify your fit and transform functions so that your vocab will contain only 50 terms with top idf scores.</li>
    <br>
    <li>This task is similar to your previous task, just that here your vocabulary is limited to only top 50 features names based on their idf values. Basically your output will have exactly 50 columns and the number of rows will depend on the number of documents you have in your corpus.</li>
    <br>
    <li>Here you will be give a pickle file, with file name <strong>cleaned_strings</strong>. You would have to load the corpus from this file and use it as input to your tfidf vectorizer.</li>
    <br>
    <li>Steps to approach this task:
    <ol>
        <li> You would have to write both fit and transform methods for your custom implementation of tfidf vectorizer, just like in the previous task. Additionally, here you have to limit the number of features generated to 50 as described above.</li>
        <li> Now sort your vocab based in descending order of idf values and print out the words in the sorted voacb after you fit your data. Here you should be getting only 50 terms in your vocab. And make sure to print idf values for each term in your vocab. </li>
        <li> Make sure the output of your implementation is a sparse matrix. Before generating the final output, you need to normalize your sparse matrix using L2 normalization. You can refer to this link https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html </li>
        <li> Now check the output of a single document in your collection of documents,  you can convert the sparse matrix related only to that document into dense matrix and print it. And this dense matrix should contain 1 row and 50 columns. </li>
        </ol>
    </li>
    <br>
   </ul>

In [1]:
# Below is the code to load the cleaned_strings pickle file provided
# Here corpus is of list type

import pickle
with open('cleaned_strings', 'rb') as f:
    pickle_corpus = pickle.load(f)
    
# printing the length of the corpus loaded
print("Number of documents in corpus = ",len(pickle_corpus))

Number of documents in corpus =  746


Here our document corpus loded from the pickle file consists of 746 documents

In [52]:
pickle_corpus[0:5] #a sample of what our file looks like

['slow moving aimless movie distressed drifting young man',
 'not sure lost flat characters audience nearly half walked',
 'attempting artiness black white clever camera angles movie disappointed became even ridiculous acting poor plot lines almost non existent',
 'little music anything speak',
 'best scene movie gerardo trying find song keeps running head']

### Fit function to find the list of all the features in our corpus

In [53]:
pickle_f = fit(pickle_corpus)
len(pickle_f)

2886

our pickle file consists of a total of 2886 features or unique words,but as per our problem statement we require only the top 50 features based on their idf scores.

So, we will sort based on the idf values

### function defined to calculate the inverse document frequence of our features

In [54]:
def vectorizer_idf(corpus,vocab):
    idf_matrix = [];features =[]  ;
    for w in tqdm(vocab):
        df = 0
        for doc in corpus:
            s = set(doc.split()) #to eliminate the duplicate word count in a document
            for i in s:
                if i == w:
                    df+=1
                    idf = 1+(math.log((1+len(corpus))/(1+df))) #idf formula mentioned above
        idf_matrix.append(idf)
        features.append(w)
    features_dict = dict(zip(features,idf_matrix)) #creating a dictionary of features & their idf scores
    features_pickle = sorted(features_dict.items(), key=lambda item: item[1],reverse=True)[:50] #sorting the top 50 features based on idf scores
    features_top = [features_idf[i][0] for i in range(len(features_idf))] #to get our top 50 features from features-idf dictionary
    return(features_pickle)

In [55]:
features_idf = vectorizer_idf(pickle_corpus,pickle_f)
features_idf

100%|█████████████████████████████████████████████████████████████████████████████| 2886/2886 [00:03<00:00, 756.09it/s]


[('aailiyah', 6.922918004572872),
 ('abandoned', 6.922918004572872),
 ('abroad', 6.922918004572872),
 ('abstruse', 6.922918004572872),
 ('academy', 6.922918004572872),
 ('accents', 6.922918004572872),
 ('accessible', 6.922918004572872),
 ('acclaimed', 6.922918004572872),
 ('accolades', 6.922918004572872),
 ('accurate', 6.922918004572872),
 ('accurately', 6.922918004572872),
 ('achille', 6.922918004572872),
 ('ackerman', 6.922918004572872),
 ('actions', 6.922918004572872),
 ('adams', 6.922918004572872),
 ('add', 6.922918004572872),
 ('added', 6.922918004572872),
 ('admins', 6.922918004572872),
 ('admiration', 6.922918004572872),
 ('admitted', 6.922918004572872),
 ('adrift', 6.922918004572872),
 ('adventure', 6.922918004572872),
 ('aesthetically', 6.922918004572872),
 ('affected', 6.922918004572872),
 ('affleck', 6.922918004572872),
 ('afternoon', 6.922918004572872),
 ('aged', 6.922918004572872),
 ('ages', 6.922918004572872),
 ('agree', 6.922918004572872),
 ('agreed', 6.922918004572872),

These are our top 50 features & their corresponding idf score values

In [24]:
features_top = [features_idf[i][0] for i in range(len(features_idf))] #top 50 features from features-idf dictionary
features_top

['aailiyah',
 'abandoned',
 'abroad',
 'abstruse',
 'academy',
 'accents',
 'accessible',
 'acclaimed',
 'accolades',
 'accurate',
 'accurately',
 'achille',
 'ackerman',
 'actions',
 'adams',
 'add',
 'added',
 'admins',
 'admiration',
 'admitted',
 'adrift',
 'adventure',
 'aesthetically',
 'affected',
 'affleck',
 'afternoon',
 'aged',
 'ages',
 'agree',
 'agreed',
 'aimless',
 'aired',
 'akasha',
 'akin',
 'alert',
 'alike',
 'allison',
 'allow',
 'allowing',
 'alongside',
 'amateurish',
 'amaze',
 'amazed',
 'amazingly',
 'amusing',
 'amust',
 'anatomist',
 'angel',
 'angela',
 'angelina']

So, these are our top 50 features based on idf score & for further processing we'll consider these features

In [25]:
tfidf_corpus = transform(pickle_corpus,features_top)

Sparse matrix: 100%|█████████████████████████████████████████████████████████████| 746/746 [00:00<00:00, 149603.19it/s]


In [26]:
tfidf_corpus 

<746x50 sparse matrix of type '<class 'numpy.float64'>'
	with 50 stored elements in Compressed Sparse Row format>

In [27]:
print(tfidf_corpus)

  (0, 30)	1.0
  (68, 24)	1.0
  (72, 29)	1.0
  (74, 31)	1.0
  (119, 33)	1.0
  (135, 3)	0.3779644730092272
  (135, 10)	0.3779644730092272
  (135, 18)	0.3779644730092272
  (135, 20)	0.3779644730092272
  (135, 36)	0.3779644730092272
  (135, 40)	0.3779644730092272
  (135, 41)	0.3779644730092272
  (176, 49)	1.0
  (181, 13)	1.0
  (192, 21)	1.0
  (193, 23)	1.0
  (216, 2)	1.0
  (222, 47)	1.0
  (225, 19)	1.0
  (227, 17)	1.0
  (241, 44)	1.0
  (270, 1)	1.0
  (290, 25)	1.0
  (333, 26)	1.0
  (334, 15)	1.0
  (341, 43)	1.0
  (344, 42)	1.0
  (348, 8)	1.0
  (377, 37)	1.0
  (409, 5)	1.0
  (430, 39)	1.0
  (457, 45)	1.0
  (461, 4)	1.0
  (465, 38)	1.0
  (475, 35)	1.0
  (493, 6)	1.0
  (500, 48)	1.0
  (548, 0)	0.7071067811865475
  (548, 32)	0.7071067811865475
  (608, 14)	1.0
  (612, 11)	1.0
  (620, 46)	1.0
  (632, 7)	1.0
  (644, 12)	0.7071067811865475
  (644, 27)	0.7071067811865475
  (664, 28)	1.0
  (667, 22)	1.0
  (691, 34)	1.0
  (697, 9)	1.0
  (722, 16)	1.0


In [31]:
doc_idx = int(input('Select the document number:->  ')) #input from user for the document number
the_doc = [pickle_corpus[doc_idx]] 
tfidf_sm = transform(the_doc,features_top) #transform function on our selected document
tfidf_dm = csr_matrix.todense(tfidf_sm) #to convert sparse matrix to dense matrix
print("Selected Document:\n",the_doc)
print("\nThe matrix for the selected document is:\n\n ",tfidf_dm)

Select the document number:->  135


Sparse matrix: 100%|███████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 1019.52it/s]

Selected Document:
 ['fact hard remember part ray charles acted not played man ray charles legendary ray charles life provided excellent biographical material film goes well beyond another movie musician hitchcock great director ironically mostly find films total waste time watch secondly hitchcock pretty much perfected thriller chase movie pandering audience sabotages films hence whole story lacks certain energy plot simply rumbles like machine desperately depending addition new scenes usual hitchcock logic flaws mishima extremely uninteresting chilly unremarkable movie author living working chilly abstruse culture flat reenactments not hold attention emotionally adrift stagy rest sits awful soldiers singing songs masculinity pledge hairsplitting purity admiration swords etc bore pieces kill momentum movie quicker anyone else schrader made resume full lousy amateurish films first watched movie loved totally fascinated music dancing everything however recently watched whole thing dvd c

Reference: countvectorizer reference notebook(AAIC), 
https://scikit-learn.org/stable/